File: scripts/audit_unused_site_files.mjs
#!/usr/bin/env node
/**
 * Audit unused files in a static site repo (conservative).
 *
 * Entrypoints:
 * - index.html (repo root)
 * - all site/(recursive)/*.html
 *
 * Parses local references from:
 * - HTML attributes: href, src, poster, data-src, data-href
 * - CSS url(...) references from linked stylesheets (local only)
 *
 * Inventories:
 * - site/, html/, files/, docs/
 * - any root directories starting with "_freeze"
 * - root: index.html, robots.txt, sitemap.xml, CNAME
 *
 * Writes:
 * - reports/unused-site-files.json
 */

import fs from "node:fs/promises";
import path from "node:path";
import process from "node:process";

const REFERENCE_ATTR_RE =
  /\b(?:href|src|poster|data-src|data-href)\s*=\s*["']([^"']+)["']/gi;
const CSS_URL_RE = /url\(\s*(['"]?)(.*?)\1\s*\)/gi;

function isExternalUrl(raw) {
  const v = String(raw ?? "").trim();
  if (!v) return true;
  if (v.startsWith("#")) return true;
  const lowered = v.toLowerCase();
  return (
    lowered.startsWith("http://") ||
    lowered.startsWith("https://") ||
    lowered.startsWith("mailto:") ||
    lowered.startsWith("tel:") ||
    lowered.startsWith("javascript:") ||
    lowered.startsWith("data:")
  );
}

function normalizeRef(raw) {
  const v = String(raw ?? "").trim();
  if (!v || isExternalUrl(v)) return null;

  // Strip query/hash via URL (supports relative URLs with a dummy base).
  let pathname = v;
  try {
    const u = new URL(v, "https://example.invalid/");
    pathname = u.pathname ?? v;
  } catch {
    // fall back to basic split
    pathname = v.split("#")[0].split("?")[0];
  }

  try {
    pathname = decodeURIComponent(pathname);
  } catch {
    // ignore bad encodings
  }

  while (pathname.startsWith("./")) pathname = pathname.slice(2);
  if (pathname.startsWith("/")) pathname = pathname.slice(1);
  if (!pathname) return null;
  if (pathname.endsWith("/")) return null;

  return pathname;
}

function toRepoRel(repoRoot, absPath) {
  const rel = path.relative(repoRoot, absPath);
  if (rel.startsWith("..") || path.isAbsolute(rel)) return null;
  return rel.split(path.sep).join("/");
}

function resolveAgainst(repoRoot, baseFileAbs, refPath) {
  const baseDir = path.dirname(baseFileAbs);
  let abs;
  if (
    refPath.startsWith("site/") ||
    refPath.startsWith("files/") ||
    refPath.startsWith("html/") ||
    refPath.startsWith("docs/")
  ) {
    abs = path.resolve(repoRoot, refPath);
  } else {
    abs = path.resolve(baseDir, refPath);
  }
  return abs;
}

async function exists(p) {
  try {
    await fs.stat(p);
    return true;
  } catch {
    return false;
  }
}

async function readText(p) {
  try {
    return await fs.readFile(p, "utf8");
  } catch {
    return "";
  }
}

function extractHtmlRefs(htmlText) {
  const refs = [];
  for (const m of htmlText.matchAll(REFERENCE_ATTR_RE)) refs.push(m[1]);
  return refs;
}

function extractCssRefs(cssText) {
  const refs = [];
  for (const m of cssText.matchAll(CSS_URL_RE)) refs.push(m[2]);
  return refs;
}

async function* walkFiles(rootAbs) {
  const stack = [rootAbs];
  while (stack.length) {
    const cur = stack.pop();
    let entries;
    try {
      entries = await fs.readdir(cur, { withFileTypes: true });
    } catch {
      continue;
    }
    for (const ent of entries) {
      const abs = path.join(cur, ent.name);
      if (ent.isDirectory()) stack.push(abs);
      else if (ent.isFile()) yield abs;
    }
  }
}

async function listInventory(repoRoot) {
  const roots = ["site", "html", "files", "docs"];
  const out = [];

  // Root singletons
  const singles = ["index.html", "robots.txt", "sitemap.xml", "CNAME"];
  for (const s of singles) {
    const abs = path.join(repoRoot, s);
    if (await exists(abs)) {
      const st = await fs.stat(abs);
      out.push({ path: s, size: st.size });
    }
  }

  // Root dirs
  for (const r of roots) {
    const absRoot = path.join(repoRoot, r);
    if (!(await exists(absRoot))) continue;
    for await (const abs of walkFiles(absRoot)) {
      const rel = toRepoRel(repoRoot, abs);
      if (!rel) continue;
      const st = await fs.stat(abs);
      out.push({ path: rel, size: st.size });
    }
  }

  // Any root directories starting with "_freeze"
  try {
    const rootEntries = await fs.readdir(repoRoot, { withFileTypes: true });
    for (const ent of rootEntries) {
      if (!ent.isDirectory()) continue;
      if (!ent.name.startsWith("_freeze")) continue;
      const absRoot = path.join(repoRoot, ent.name);
      for await (const abs of walkFiles(absRoot)) {
        const rel = toRepoRel(repoRoot, abs);
        if (!rel) continue;
        const st = await fs.stat(abs);
        out.push({ path: rel, size: st.size });
      }
    }
  } catch {
    // ignore
  }

  // De-dupe (same file might be included twice if overlap)
  const seen = new Set();
  const deduped = [];
  for (const fi of out) {
    if (seen.has(fi.path)) continue;
    seen.add(fi.path);
    deduped.push(fi);
  }
  return deduped;
}

async function main() {
  const args = process.argv.slice(2);
  const getArg = (name, def) => {
    const idx = args.indexOf(name);
    if (idx === -1) return def;
    return args[idx + 1] ?? def;
  };

  const repoRoot = path.resolve(getArg("--root", "."));
  const reportRel = getArg("--report", "reports/unused-site-files.json");
  const reportAbs = path.resolve(repoRoot, reportRel);

  const entrypoints = [];
  const rootIndex = path.join(repoRoot, "index.html");
  if (await exists(rootIndex)) entrypoints.push(rootIndex);
  const siteDir = path.join(repoRoot, "site");
  if (await exists(siteDir)) {
    for await (const abs of walkFiles(siteDir)) {
      if (abs.toLowerCase().endsWith(".html")) entrypoints.push(abs);
    }
  }

  const referencedFiles = new Set();
  const referencedMissing = new Set();
  const cssToParse = new Set();

  for (const htmlAbs of entrypoints) {
    const htmlText = await readText(htmlAbs);
    for (const raw of extractHtmlRefs(htmlText)) {
      const norm = normalizeRef(raw);
      if (!norm) continue;
      const resolvedAbs = resolveAgainst(repoRoot, htmlAbs, norm);
      const rel = toRepoRel(repoRoot, resolvedAbs);
      if (!rel) continue;
      if (await exists(resolvedAbs)) referencedFiles.add(rel);
      else referencedMissing.add(rel);
      if (rel.toLowerCase().endsWith(".css") && (await exists(resolvedAbs))) {
        cssToParse.add(resolvedAbs);
      }
    }
  }

  for (const cssAbs of Array.from(cssToParse).sort()) {
    const cssText = await readText(cssAbs);
    for (const raw of extractCssRefs(cssText)) {
      const norm = normalizeRef(raw);
      if (!norm) continue;
      const resolvedAbs = resolveAgainst(repoRoot, cssAbs, norm);
      const rel = toRepoRel(repoRoot, resolvedAbs);
      if (!rel) continue;
      if (await exists(resolvedAbs)) referencedFiles.add(rel);
      else referencedMissing.add(rel);
    }
  }

  const inventory = await listInventory(repoRoot);
  const referencedSet = referencedFiles;

  const allowlistExact = new Set([
    "site/search.json",
    "site/listings.json",
    "site/sitemap.xml",
    "site/robots.txt",
  ]);

  const unusedCandidates = [];
  for (const fi of inventory) {
    if (allowlistExact.has(fi.path)) continue;
    if (referencedSet.has(fi.path)) continue;
    if (fi.path === "index.html") continue;
    if (fi.path.startsWith("site/") && fi.path.toLowerCase().endsWith(".html")) continue;
    unusedCandidates.push(fi);
  }

  const safeDelete = [];
  for (const fi of unusedCandidates) {
    const p = fi.path;
    if (p.startsWith("docs/")) safeDelete.push(fi);
    else if (p.startsWith("_freeze")) safeDelete.push(fi);
    else if (p.startsWith("site/") && p.toLowerCase().endsWith(".ipynb")) safeDelete.push(fi);
    else if (p.startsWith("site/") && p.toLowerCase().endsWith(".map")) safeDelete.push(fi);
  }

  await fs.mkdir(path.dirname(reportAbs), { recursive: true });
  const sum = (arr) => arr.reduce((a, b) => a + (b.size ?? 0), 0);

  const report = {
    entrypointsCount: entrypoints.length,
    referencedFilesCount: referencedFiles.size,
    referencedMissingCount: referencedMissing.size,
    referencedMissing: Array.from(referencedMissing).sort(),
    inventoryFilesCount: inventory.length,
    unusedCandidatesCount: unusedCandidates.length,
    unusedCandidatesTotalBytes: sum(unusedCandidates),
    unusedCandidates: unusedCandidates.map((x) => ({ path: x.path, size: x.size })).sort((a, b) => a.path.localeCompare(b.path)),
    safeDeleteCount: safeDelete.length,
    safeDeleteTotalBytes: sum(safeDelete),
    safeDelete: safeDelete.map((x) => ({ path: x.path, size: x.size })).sort((a, b) => a.path.localeCompare(b.path)),
    notes: [
      "This audit is conservative. Anything ambiguous is treated as referenced.",
      "Review safeDelete list; only those are intended for automatic deletion in safe mode.",
    ],
  };

  await fs.writeFile(reportAbs, JSON.stringify(report, null, 2) + "\n", "utf8");
  console.log(`Wrote report: ${reportAbs}`);
  console.log(`Entrypoints: ${entrypoints.length}`);
  console.log(`Inventory files: ${inventory.length}`);
  console.log(`Unused candidates: ${unusedCandidates.length} (${report.unusedCandidatesTotalBytes} bytes)`);
  console.log(`Safe delete: ${safeDelete.length} (${report.safeDeleteTotalBytes} bytes)`);
}

await main();

File: scripts/audit_unused_site_files.py
#!/usr/bin/env python3
"""
Audit unused files in a static site repo (conservative).

Entrypoints:
- index.html (repo root)
- all site/**/*.html

We parse local references from:
- HTML attributes: href, src, poster, data-src, data-href
- CSS url(...) references from linked stylesheets (local only)

We then inventory on-disk files under:
- site/
- html/
- files/
- docs/
- root: index.html, robots.txt, sitemap.xml, CNAME

Output:
- reports/unused-site-files.json

This script is intentionally conservative: anything ambiguous is treated as
referenced (i.e., NOT a deletion candidate).
"""

from __future__ import annotations

import argparse
import json
import os
import re
import sys
from dataclasses import dataclass
from html.parser import HTMLParser
from pathlib import Path
from typing import Iterable, Optional
from urllib.parse import unquote, urlsplit


REFERENCE_ATTRS = {"href", "src", "poster", "data-src", "data-href"}
CSS_URL_RE = re.compile(r"url\(\s*(['\"]?)(.*?)\1\s*\)", re.IGNORECASE)


def is_external_url(raw: str) -> bool:
    v = raw.strip()
    if not v:
        return True
    if v.startswith("#"):
        return True
    # Schemes and special pseudo-URLs we never treat as local files
    lowered = v.lower()
    return lowered.startswith((
        "http://",
        "https://",
        "mailto:",
        "tel:",
        "javascript:",
        "data:",
    ))


def normalize_ref(raw: str) -> Optional[str]:
    """
    Convert a URL-ish reference into a normalized repo-relative path string.
    Returns None if not a local path.
    """
    raw = raw.strip()
    if not raw or is_external_url(raw):
        return None

    parts = urlsplit(raw)
    if parts.scheme or parts.netloc:
        return None

    path = unquote(parts.path)
    if not path or path.endswith("/"):
        # We don't try to infer directory index files; treat as non-file reference.
        return None

    # Strip leading "./"
    while path.startswith("./"):
        path = path[2:]

    # Strip leading "/" (site is served from repo root)
    if path.startswith("/"):
        path = path[1:]

    if not path:
        return None

    return path


def resolve_against(base_file: Path, ref_path: str, repo_root: Path) -> Optional[Path]:
    """
    Resolve a normalized ref path to a concrete filesystem path.
    If ref_path is relative, resolve from base_file.parent; if it's already root-like,
    resolve from repo_root.
    """
    # If it contains parent traversal, we still resolve it but normalize.
    if ref_path.startswith(("site/", "files/", "html/", "docs/")) or "/" in ref_path[:1]:
        cand = (repo_root / ref_path).resolve()
    else:
        cand = (base_file.parent / ref_path).resolve()

    try:
        return cand.relative_to(repo_root.resolve()) and cand
    except Exception:
        # Points outside repo; ignore.
        return None


class RefHTMLParser(HTMLParser):
    def __init__(self) -> None:
        super().__init__(convert_charrefs=True)
        self.refs: list[str] = []

    def handle_starttag(self, tag: str, attrs: list[tuple[str, Optional[str]]]) -> None:
        for k, v in attrs:
            if k in REFERENCE_ATTRS and v:
                self.refs.append(v)


def extract_html_refs(html_text: str) -> list[str]:
    p = RefHTMLParser()
    try:
        p.feed(html_text)
    except Exception:
        # Be conservative: if parsing fails, return empty; we will not delete uncertain files anyway.
        return []
    return p.refs


def extract_css_refs(css_text: str) -> list[str]:
    refs: list[str] = []
    for m in CSS_URL_RE.finditer(css_text):
        refs.append(m.group(2))
    return refs


@dataclass(frozen=True)
class FileInfo:
    path: str
    size: int


def iter_files(repo_root: Path, rel_dirs: Iterable[str]) -> list[FileInfo]:
    out: list[FileInfo] = []
    for d in rel_dirs:
        p = repo_root / d
        if not p.exists():
            continue
        if p.is_file():
            st = p.stat()
            out.append(FileInfo(path=d.replace("\\", "/"), size=st.st_size))
            continue
        for f in p.rglob("*"):
            if f.is_file():
                st = f.stat()
                rel = f.relative_to(repo_root).as_posix()
                out.append(FileInfo(path=rel, size=st.st_size))
    return out


def main() -> int:
    ap = argparse.ArgumentParser()
    ap.add_argument("--root", default=".", help="Repo root (default: .)")
    ap.add_argument("--report", default="reports/unused-site-files.json")
    args = ap.parse_args()

    repo_root = Path(args.root).resolve()

    # Entrypoints
    entrypoints: list[Path] = []
    root_index = repo_root / "index.html"
    if root_index.exists():
        entrypoints.append(root_index)
    entrypoints.extend(sorted((repo_root / "site").rglob("*.html")) if (repo_root / "site").exists() else [])

    referenced_files: set[str] = set()
    referenced_missing: set[str] = set()

    # We also track linked CSS to parse url(...) references.
    css_to_parse: set[Path] = set()

    for html_file in entrypoints:
        try:
            html_text = html_file.read_text(encoding="utf-8", errors="ignore")
        except Exception:
            continue

        raw_refs = extract_html_refs(html_text)
        for raw in raw_refs:
            norm = normalize_ref(raw)
            if norm is None:
                continue
            resolved = resolve_against(html_file, norm, repo_root)
            if resolved is None:
                continue
            rel = resolved.relative_to(repo_root).as_posix()

            # Track local refs even if file doesn't exist (helps find broken links)
            if resolved.exists():
                referenced_files.add(rel)
            else:
                referenced_missing.add(rel)

            # Collect local CSS for url(...) parsing
            if rel.lower().endswith(".css") and resolved.exists():
                css_to_parse.add(resolved)

    # Parse CSS url(...) references
    for css_file in sorted(css_to_parse):
        try:
            css_text = css_file.read_text(encoding="utf-8", errors="ignore")
        except Exception:
            continue
        for raw in extract_css_refs(css_text):
            norm = normalize_ref(raw)
            if norm is None:
                continue
            resolved = resolve_against(css_file, norm, repo_root)
            if resolved is None:
                continue
            rel = resolved.relative_to(repo_root).as_posix()
            if resolved.exists():
                referenced_files.add(rel)
            else:
                referenced_missing.add(rel)

    # Inventory (web-facing)
    inventory_dirs = [
        "site",
        "html",
        "files",
        "docs",
        "robots.txt",
        "sitemap.xml",
        "CNAME",
        "index.html",
    ]
    all_files = iter_files(repo_root, inventory_dirs)
    all_paths = {fi.path for fi in all_files}

    # Allowlist things that may be fetched dynamically even if not referenced
    allowlist_exact = {
        "site/search.json",
        "site/listings.json",
        "site/sitemap.xml",
        "site/robots.txt",
    }

    unused = []
    for fi in all_files:
        if fi.path in allowlist_exact:
            continue
        if fi.path in referenced_files:
            continue
        # Don't delete entrypoint HTMLs even if unreferenced by others.
        if fi.path == "index.html" or (fi.path.startswith("site/") and fi.path.lower().endswith(".html")):
            continue
        unused.append(fi)

    # High-confidence safe-to-delete subset (very conservative)
    safe_delete: list[FileInfo] = []
    for fi in unused:
        p = fi.path
        # Old accidental cache folder in this repo (has a space); safe to delete when unreferenced
        if p.startswith("_freeze "):
            safe_delete.append(fi)
            continue
        # Docs folder not used when Pages source is root (still only if unreferenced)
        if p.startswith("docs/"):
            safe_delete.append(fi)
            continue
        # Notebook artifacts in published site output
        if p.startswith("site/") and p.lower().endswith(".ipynb"):
            safe_delete.append(fi)
            continue
        # Source maps (if any) are safe to remove when unreferenced
        if p.startswith("site/") and p.lower().endswith(".map"):
            safe_delete.append(fi)
            continue

    report_path = (repo_root / args.report).resolve()
    report_path.parent.mkdir(parents=True, exist_ok=True)

    def pack(infos: list[FileInfo]) -> list[dict]:
        return [{"path": x.path, "size": x.size} for x in sorted(infos, key=lambda z: (z.path))]

    out = {
        "entrypointsCount": len(entrypoints),
        "referencedFilesCount": len(referenced_files),
        "referencedMissingCount": len(referenced_missing),
        "referencedMissing": sorted(referenced_missing),
        "inventoryFilesCount": len(all_files),
        "unusedCandidatesCount": len(unused),
        "unusedCandidatesTotalBytes": sum(x.size for x in unused),
        "unusedCandidates": pack(unused),
        "safeDeleteCount": len(safe_delete),
        "safeDeleteTotalBytes": sum(x.size for x in safe_delete),
        "safeDelete": pack(safe_delete),
        "notes": [
            "This audit is conservative. Anything ambiguous is treated as referenced.",
            "Review safeDelete list; only those are intended for automatic deletion in safe mode.",
        ],
    }
    report_path.write_text(json.dumps(out, indent=2, sort_keys=True) + "\n", encoding="utf-8")

    # Console summary
    print(f"Wrote report: {report_path}")
    print(f"Entrypoints: {len(entrypoints)}")
    print(f"Inventory files: {len(all_files)}")
    print(f"Unused candidates: {len(unused)} ({out['unusedCandidatesTotalBytes']} bytes)")
    print(f"Safe delete: {len(safe_delete)} ({out['safeDeleteTotalBytes']} bytes)")

    # Non-zero exit if broken refs found (useful for CI), but keep it soft for now.
    return 0


if __name__ == "__main__":
    raise SystemExit(main())



